ingester/ingester.go

```diff
@@ -487,42 +489,70 @@ func (i *Ingester) flushAllUsers(immediate bool) {
 	i.userStateLock.Unlock()
 
 	for id, state := range userState {
-		i.flushUser(id, state, immediate)
-	}
-}
-
-func (i *Ingester) flushUser(userID string, userState *userState, immediate bool) {
-	for pair := range userState.fpToSeries.iter() {
-		i.flushSeries(userState, pair.fp, pair.series, immediate)
+		for pair := range state.fpToSeries.iter() {
+			state.fpLocker.Lock(pair.fp)
+			i.sweepSeries(id, pair.fp, pair.series, immediate)
+			state.fpLocker.Unlock(pair.fp)
+		}
 	}
 
-	// TODO: this is probably slow, and could be done in a better way.
 	i.userStateLock.Lock()
-	if userState.fpToSeries.length() == 0 {
-		delete(i.userState, userID)
+	for id, state := range userState {
+		if state.fpToSeries.length() == 0 {
+			delete(i.userState, id)
+		}
```

Comment:
> If code somewhere else decided to change an existing i.userState entry, this would become a race.

It would.  But no such code exists.

The race here is between calls to `getStateFor` (which gets the userState) and `getOrCreateSeries` (which gets or creates a series).  Currently we might delete the `userState` in the middle of these two calls, which will cause some samples to be dropped.

Its a bit tricky to fix without hurting performance, and as this isn't a regression, I'll file a separate issues and work on fixing that next.  Sound good?